pkg/leaderelection/context.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leaderelection
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+
+	"knative.dev/pkg/controller"
+	pkgleaderelection "knative.dev/pkg/leaderelection"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/system"
+)
```

Comment:
I just came across this trying to enable HA by default for our controlplane components, which eliminates the `enabledComponents` and the `LeaderElect` fields from knative/pkg.

This code has zero eventing dependencies, and forks a ton of the PKG logic into eventing. 😞 

Is there a design or a feature track document for this anywhere...?

Reply:
this code is almost the same as the one found in PKG except it depends on adapter.Adapter instead of reconciler.LeaderAware. (there was issues with directly imported adapter.Adapter from `context.go` so the interface is duplicated in this file).

I would like to consolidate the two together, and maybe `reconciler.LeaderAware` is the right abstraction but I'm not sure yet. That's on my TODO list for 0.17.